### ISSUE TITLE
Autocomplete returns all variables

### ISSUE THREAD

**Issue by Kaali** [ananthakumaran/tide#208]
When I use autocompletion with `company-tide`, I see all the variables in the file as candidates. Most of them have no annotations. I tried to debug it a little, and noticed that the server is returning a lot of names with kind "warning". When I filter those out from the candidates, the results look as I would expect them to look.

I cannot send any code to replicate the issue, but I can help debugging the issue if someone tells me how to see the raw server data.

**Comment by ananthakumaran** [comment]
look at https://github.com/ananthakumaran/tide#debugging section. You can see redirect all the logs from tsserver to a file.

**Comment by Kaali** [comment]
Thanks. tsserver does return kind warning names on completion. Is this expected behavior? The names are completely unrelated to the type which members I'm autocompleting on.

**Comment by ananthakumaran** [comment]
can you share some part of the logs? probably it's something they have added

**Comment by josteink** [comment]
If you can create a minimum POC repo where the issue can be reproduced, that would be of great help.

**Comment by Kaali** [comment]
I tried to create a small project to repeat the problem. But I cannot reproduce it without including my customers proprietary code. So it's that code that does something, which might mean there is a bug in tsserver. tide seems to work fine in a small project, but on the customer project it dumps out all the names accessible in the scope as completion options.

I can get a log from that, but I doubt it helps.

**Comment by Kaali** [comment]
I'm not running tide with this change to fix the problem:

`(defun tide-annotate-completions (completions prefix file-location)
  (-map
   (lambda (completion)
     (let ((name (plist-get completion :name)))
       (put-text-property 0 1 'file-location file-location name)
       (put-text-property 0 1 'completion completion name)
       name))
   (-sort
    'tide-compare-completions
    (-filter
     (lambda (completion)
       ;; (string-prefix-p prefix (plist-get completion :name)))
       (and (string-prefix-p prefix (plist-get completion :name))
            (not (equal (plist-get completion :kind) "warning"))))
     completions))))
`

**Comment by Nek** [comment]
Hello @Kaali 

I'm new to Emacs and can't apply the fix myself. How do I use it?

**Comment by josteink** [comment]
Copy it to the scratch buffer and M-x eval-buffer.

**Comment by ananthakumaran** [comment]
It would be helpful if you could come up with a minimal code that could reproduce the issue. As of now, we are not sure whether the problem lies in tide or tsserver.

**Comment by Nek** [comment]
Got it. I'll try to do it. Though, I'm not sure what minimal code is in this case. I'd be grateful if you could elaborate what you expect to see, maybe give me an example.

**Comment by ananthakumaran** [comment]
Right now we can't reproduce the issue, so we need to have minimal typescript project that has this problem. Also from @Kaali response, it seems most likely a issue in tsserver itself, not in tide.

**Comment by Nek** [comment]
It's clear now. Thanks.

**Comment by Nek** [comment]
This gives me an error. I don't use TS, so, it's just JS.
https://github.com/Nek/tide-error-demo

**Comment by Nek** [comment]
Just for the record, there is another problem I've noticed while patching the error locally.

<img width="412" alt="screen shot 2018-05-19 at 6 56 30 pm" src="https://user-images.githubusercontent.com/132011/40270952-9adfe172-5b96-11e8-80d9-83c6bf8b4226.png">

As you can see, there are some duplicates of **delay**. When I've checked `candidates` I could see the duplication of entries and filtered the resulting completions list using `-distinct`.

**Comment by ananthakumaran** [comment]
The problem with javascript has already been brought up. see #143, you can set `tide-sort-completions-by-kind` to true to get somewhat better sorting.

**Comment by Nek** [comment]
It's already set to true.

<img width="852" alt="screen shot 2018-05-19 at 7 04 18 pm" src="https://user-images.githubusercontent.com/132011/40270999-7a4c4c9c-5b97-11e8-9fbd-3015d5922043.png">

My tide's configuration. Nothing else is customized outside of it:
```
(use-package tide
  :delight ".ts "
  :ensure t
  :bind
  (("M-s M-r" . tide-references))
  :init
  (setq tide-always-show-documentation t)
  (setq tide-completion-detailed nil)
  (setq tide-completion-enable-autoimport-suggestions t)
  (setq tide-completion-ignore-case t)
  (setq tide-default-mode "JS")
  (setq tide-format-options nil)
  (setq tide-sort-completions-by-kind t)
  (setq tide-sync-request-timeout 0)
  :config
  (setq tide-tsserver-executable "/usr/local/bin/tsserver")
  (setq company-tooltip-align-annotations t)
  (flycheck-add-next-checker 'javascript-eslint 'javascript-tide 'append)
  (flycheck-add-next-checker 'javascript-eslint 'jsx-tide 'append)
  :hook
  (js2-mode . setup-tide-mode)
  (rjsx-mode . setup-tide-mode))
```

**Comment by Nek** [comment]
Another thing to mention.

![tidebug](https://user-images.githubusercontent.com/132011/40271214-7b259940-5b9a-11e8-9eb7-a360f130dbcc.gif)

The logged string is `:kind` of candidate passed to company-mode. As you can see it's "warning" and not something useful.

**Comment by vshender** [comment]
It seems that just skipping entries with "warning" kind solves the problem. BTW, vim plugin uses this approach: https://github.com/Quramy/tsuquyomi/blob/master/autoload/tsuquyomi.vim#L387.

**Comment by Nek** [comment]
Thanks for the additional info. I patched tide locally and I'm doing exactly the filtering. I also filter duplicate entries as I've mentioned above.

```
(defun tide-annotate-completions
    (completions prefix file-location)
    (-map (lambda (completion)
            (let ((name (plist-get completion
                          :name)))
              (put-text-property 0 1 'file-location
                file-location name)
              (put-text-property 0 1 'completion
                completion name)
              (put-text-property 0 1 'prefix prefix
                name) name))
      (let ((filtered (-filter (lambda (completion)
                                 (and (string-prefix-p
                                        prefix
                                        (plist-get
                                          completion
                                          :name))
                                   (not (equal
                                          (plist-get
                                            completion
                                            :kind)
                                          "warning" ) ))) completions)))
        (-distinct (if tide-sort-completions-by-kind
                     (-sort 'tide-compare-completions
                       filtered) filtered)))))
```

**Comment by ananthakumaran** [comment]
@Nek can you open a PR for filtering `warning`. Put it behind a config flag and set it to `nil` by default and only do it for non typescript files (check if file extension is `.ts` or `.tsx`).

Do you get duplicate entries after filtering warnings? Some duplicates might be valid, like import suggestion from two different packages etc.

**Comment by vshender** [comment]
@ananthakumaran I've already done this in my fork. I'm going to add some tests and cleanups today and open PR.

**Comment by vshender** [comment]
@ananthakumaran @Nek https://github.com/ananthakumaran/tide/pull/257

**Comment by ananthakumaran** [comment]
fixed by #257
